lib/msal-common/src/response/ResponseHandler.ts

```diff
@@ -0,0 +1,300 @@
+        authority: Authority
+    ): AuthenticationResult {
+        const authenticationResult: AuthenticationResult = {
+            uniqueId: "",
+            tenantId: "",
+            tokenType: "",
+            idToken: null,
+            idTokenClaims: null,
+            accessToken: "",
+            refreshToken: "",
+            scopes: [],
+            expiresOn: null,
+            familyId: null,
+        };
+
+        // IdToken
+        const idTokenObj = new IdToken(
+            serverTokenResponse.id_token,
+            this.cryptoObj
+        );
+
+        // if account is not in cache, append it to the cache
+        this.addAccountToCache(serverTokenResponse, idTokenObj, authority);
+
+        // TODO: Check how this changes for auth code response
+        const expiresSeconds = Number(idTokenObj.claims.exp);
+        if (expiresSeconds && !authenticationResult.expiresOn) {
+            authenticationResult.expiresOn = new Date(expiresSeconds * 1000);
+        }
+
+        // Expiration calculation
+        const expiresInSeconds =
+            TimeUtils.nowSeconds() + serverTokenResponse.expires_in;
+        const extendedExpiresInSeconds =
+            expiresInSeconds + serverTokenResponse.ext_expires_in;
+        // Set consented scopes in response
+        const responseScopes = ScopeSet.fromString(
+            serverTokenResponse.scope,
+            this.clientId,
+            true
+        );
+
+        return {
+            ...authenticationResult,
+            uniqueId: idTokenObj.claims.oid || idTokenObj.claims.sub,
+            tenantId: idTokenObj.claims.tid,
+            idToken: idTokenObj.rawIdToken,
+            idTokenClaims: idTokenObj.claims,
+            accessToken: serverTokenResponse.access_token,
+            refreshToken: serverTokenResponse.refresh_token,
+            expiresOn: new Date(expiresInSeconds),
+            extExpiresOn: new Date(extendedExpiresInSeconds),
+            scopes: responseScopes.asArray(),
+            familyId: serverTokenResponse.foci,
+        };
+    }
+
+    /**
+     * if Account is not in the cache, generateAccount and append it to the cache
+     * @param serverTokenResponse
+     * @param idToken
+     * @param authority
+     */
+    addAccountToCache(
+        serverTokenResponse: ServerAuthorizationTokenResponse,
+        idToken: IdToken,
+        authority: Authority
+    ) {
+        const environment =
+            authority.canonicalAuthorityUrlComponents.HostNameAndPort;
+        let accountEntity: AccountEntity;
+        const cachedAccount: AccountEntity = this.uCacheManager.getAccount(
+            this.homeAccountIdentifier,
+            environment,
+            idToken.claims.tid
+        );
+        if (!cachedAccount) {
+            accountEntity = this.generateAccountEntity(
+                serverTokenResponse,
+                idToken,
+                authority
+            );
+            this.uCacheManager.addAccountEntity(accountEntity);
```

Comment:
accountEntity is not initialized if there is a cachedAccount. Is this expected behavior?

Reply:
no. Will add the logic in `silent` flow support.